Remove daemon, improve tests, add tests, fix bugs

dankwart-de opened the pull request.

vinjana commented on src/main/groovy/de/dkfz/roddy/execution/jobs/JobManagerOptions.groovy:
```diff
@@ -84,6 +91,7 @@ class JobManagerOptionsBuilder {
     JobManagerOptionsBuilder() {
         trackOnlyStartedJobs = false
         updateInterval = Duration.ofMinutes(5)
+        maxAgeOfJobsForJobQueries = Duration.ofDays(14)
```
maxTerminatedJobTrackingTime?

vinjana commented on src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy:
```diff
@@ -29,33 +30,38 @@ import java.time.format.DateTimeFormatter
-            Map<String, Object> jobDetails = runBjobs([id]).get(id)
-            queriedExtendedStates.put(id, queryJobInfo(jobDetails))
+            Map<String, Object> jobDetails = runBjobs([id], true).get(id)
+            queriedExtendedStates.put(id, convertJobDetailsMapToGenericJobInfoobject(jobDetails))
```
o -> O

vinjana commented on src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy:
```diff
@@ -109,55 +116,75 @@ class LSFJobManager extends AbstractLSFJobManager {
+            String finishTime = record["FINISH_TIME"]
+            boolean youngEnough = true
+            if (finishTime) {
+                String timeString = "${finishTime[0..-3]}"
```
L = local time?

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManagerSpec.groovy:
```diff
@@ -16,182 +16,85 @@ import de.dkfz.roddy.execution.jobs.JobState
-}
-'''
 
     void "queryJobInfo, bjobs JSON output with lists  "() {
```
change name

vinjana commented on src/test/resources/de/dkfz/roddy/execution/jobs/cluster/lsf/queryExtendedJobStateByIdTest.json:
```diff
@@ -0,0 +1,41 @@
+      "EXEC_HOST": "exec-host:exec-host",
+      "SUBMIT_TIME": "Dec 28 19:56",
+      "START_TIME": "Dec 28 19:56",
+      "FINISH_TIME": "Dec 28 19:56 L",
```
https://www.ibm.com/support/knowledgecenter/en/SSETD4_9.1.2/lsf_command_ref/bjobs.1.html

L is status information.

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManagerSpec.groovy:
```diff
@@ -16,182 +16,85 @@ import de.dkfz.roddy.execution.jobs.JobState
+
+        where:
+        resourceFile                                     | expectedJobId | expectedTool
+        "queryExtendedJobStateByIdTest.json"             | "22005"       | new File("ls -l")
```
as discussed and already in the old code: not File, but string of a command

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManagerSpec.groovy:
```diff
@@ -16,182 +16,85 @@ import de.dkfz.roddy.execution.jobs.JobState
-        List records = (List) parsedJson.getAt("RECORDS")
+    @CompileStatic
+    String localDateTimeToLSFString(LocalDateTime date) {
+        DateTimeFormatter.ofPattern('MMM ppd HH:mm').withLocale(Locale.ENGLISH).format(date) + " L"
```
This correctly specifies that this is a LSF string, but note, that not all of these use "L" and some have no such suffix. Therefore, document the L (or whatever alternative is appropriate) and maybe rename to narrow down the type of LSF date-time string that is produced.

Commit:
Big rework of test structure, remove update daemon

Update the dependency to RoddyToolLib to 2.1.1

// Library classes

Renamed GenericJobInfo to ExtendedJobInfo

Make both JobInfo classes use the @EqualsAndHashCode annotation to
improve testability of code with (Extended)JobInfo objects.

In ExtendedJobInfo:
- Reordered the variable in attempt to group them better.
- Renamed some of the variables to better match their function.
- Make the class extend JobInfo and move jobID, endTime and jobState to
  the new class.

Removed tracking and daemon fields from all classes. The daemon is now a
conceptual idea which can be implemented by software using BE.

In BatchEuphoriaJobManager:
- Unify get(Extended)JobInfo method layout.
- Implemented most methods in the class and made them final, add tests
  for them.

In ClusterBasedJobManager:
- Made withCaughtAndLoggedException public.
- Added the toJobID(), safelyParseColonSeparatedDuration(),
  safelyParseTime() and the abstract parseTime() methods.

In GridEngineBasedJobManager:
- Add parseTime() method
- Rename queryJobStates() to queryJobInfo()
- Extract the assembleJobInfoQueryCommand() method from queryJobInfo()
  The method will assemble "qstat [-u] [id] [id] [...]" by default and
  is overridden in SGEJobManager.
- Removed processQstatOutputFromPlainText(), we only parse from xml, so
  this was redundant and actually never used.
- Add safe cast and convert methods for BufferValue, Integer, Duration
  TimeUnit and others. They will safely convert/case a value from a
  String.
- Moved the processQstatOutputFromXML() method to PBSJobManager,
  SGE has a completely different implementation!

In SGEJobManager:
- Override assembleJobInfoQueryCommand()
- Add convertJobInfoMapToExtendedMap()
- Override queryExtendedJobInfo(), this works in a multi step process,
  first it will call query(All)JobInfo() to get basic information about
  jobs. Why? Because SGE provides you with different output depending
  whether active jobs can be found or not. Detailed explanations are in
  the method. After the first step, extended information are queried and
  processed.
- Add parseGPathJobInfoElement(), which will parse a job/element entry
  in extended qstat xml output
- Add extractAndParseTime() and extractStatisticsValue()

In PBSJobManager:
- Override queryExtendedJobInfo()
- Add processQstatOutputFromXMLNodeChild(), which will parse a job entry
  in the extended qstat xml output. Actually it was moved here from
  GridEngineBasedJobManager. Also the contents are reordered to match
  the corresponding methods in LSFJobManager and SGEJobManager

In LSFJobManager:
- Reduce the returned JSON result in queryJobInfo() to the most
  essential fields: jobid, job_name, stat and finish_time.
- Improve the parseTime method to cover several versions of possible
  LSF datetime reports. Detailed explanations are in the method comments
- Improve stripAwayStatusInfo()
- Remove queryExtendedJobStateById(), its in BEJobManager
- Override queryJobInfo() and queryExtendedJobInfo(), remove
  queryJobStates
- Remove filterJobMapByAge(). This is a method for a daemon and not
  covered by BatchEuphoria anymore.
- Rework convertJobDetailsMapToGenericJobInfoObject() further
- Remove safelyParseColonSeparatedDuration() and safelyParseTime(), they
  are in a super class.

In LSFRestJobManager:
- Override queryExtendedJobInfo()
- Just saw, that a lot of work is still necessary in this class. But
  please review the rest first.

// Test classes and files

Add a lot of resource files, they are mostly extracted from tests and
this is done to make the test layout more structured and the tests more
readable. To load a resource file, utilize the new BETestBaseSpec class,
which has the getResourceFile() method and a test for it.

Created the JobManagerImplementationBaseSpec class, which is now a base
specification for all job manager specific Spock specs:
- The method createJobManagerWithModifiedExecutionService() can be used
  to create a job manager, which will use the passed resource for
  queryJobInfo() or queryExtendedJobInfo() test cases. It will create
  a job manager of type T! T needs to be set, when the Spec is extended
- The method createJobManager() will create a very basic job manager of
  type T
- Also there is a list of possible methods / test features, which are
  meant to bring some basic identical structure to all job manager tests
- The class is tested in JobManagerImplementationBaseSpecSpec

Create the BatchEuphoriaJobManagerSpec:
- Make it implement JobManagerImplementationBaseSpec
- Add tests for the different final query methods in
  BatchEuphoriaJobManager

In GridEngineBasedJobManagerSpec, extend test for getExecutionHosts()

In ClusterJobManagerSpec:
- Add tests for parseColonSeparatedHHMMSSDuration()
- Fix prepareClassLogger()
- Rename tests

Add the SGEJobManagerSpec:
- Has stubs for most of the tests in JobManagerImplementationBaseSpec
- Has several tests for query(Extended)JobInfo() which try to tackle
  the slightly special behaviour of queryExtendedJobInfo()

In PBSJobManagerSpec:
- Remove all the XML variables and move content to resource files.
- Add stubs for tests in JobManagerImplementationBaseSpec
- Make tests use the new equals method of (Extended)JobInfo.
- Improved layout and readability of some tests.
- Move the parseColonSeparatedHHMMSSDuration() tests to
  ClusterJobManagerSpec

In PBSCommandParserTest:
- Reduce code by utilizing the TestExecutionService for the
  PBSJobManager.

In LSFJobManagerSpec:
- Remove the getResourceFile() method. Its in a super class.
- Make the class extend JobManagerImplementationBaseSpec and create
  test stubs
- Spockify the parseTime() test and cover all cases and also
  malformatted strings
- Make the query(E)JI() tests use the new equals feature of JobInfo
- Remove the filterJobMapByAge test

Create a test class LSFRestJobManagerSpec. It is not implemented yet!

vinjana commented on src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/sge/SGEJobManager.groovy:
```diff
@@ -134,4 +131,157 @@ class SGEJobManager extends GridEngineBasedJobManager<SGECommand> {
+        // jobInfo.queue                                                   // Info is not available
+
+        // Getting cores and nodes is a bit tricks. Nodes are not in the xml, cores via JB_pe_range
+        String requestedCoresRaw = job["JB_pe_range"]["ranges"]["RN_min"]?.toString()
```
Can we omit the `requested` prefixes?

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/BatchEuphoriaJobManagerSpec.groovy:
```diff
@@ -0,0 +1,174 @@
+        "999"  | UNKNOWN
+        "1000" | RUNNING
+        "1001" | SUSPENDED
+        "1002" | COMPLETED_SUCCESSFUL
```
Should we actually also start fixing the states? Maybe soon. The states currently code two pieces of information in the name, namely status according to batch processing system (RUNNING, UNKOWN, COMPLETED) and exit code (SUCCESSFUL, FAILURE). We should absorb the exit code into a field.

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy:
```diff
@@ -6,326 +6,245 @@
+        expected                                                | mem  | cores | nodes | walltime
+        "-l mem=1024M -l walltime=00:01:00:00 -l nodes=1:ppn=2" | 1024 | 2     | 1     | "h"
+        "-l walltime=00:01:00:00 -l nodes=1:ppn=1"              | null | null  | 1     | "h"
+        "-l mem=1024M -l nodes=1:ppn=2"                         | 1024 | 2     | null  | null
```
nodes == null? nul means use the default, right?

For clarity I'd prefer to see the inputs on the left and the expect on the right (but that is nitpicking :-P)

vinjana commented on src/test/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManagerSpec.groovy:
```diff
@@ -6,326 +6,245 @@
+
+        def jobManager = createJobManagerWithModifiedExecutionService("processExtendedQstatXMLOutputWithQueuedJob.xml")
+
+        ExtendedJobInfo expected = getGenericJobInfoObjectForTests()
```
getExtendedJobInfoobjectForTests()

vinjana commented on src/test/resources/de/dkfz/roddy/execution/jobs/cluster/sge/queryExtendedJobStatesWithQueuedJobs.xml:
```diff
@@ -0,0 +1,632 @@
+                </job_sublist>
+                <job_sublist>
+                    <VA_variable>CONFIG_FILE</VA_variable>
+                    <VA_value>/home/heinold/temp/roddyLocalTest/testproject/rpp/stds/roddyExecutionStore/exec_190204_103439242_heinold_testCorrect2/RoddyTest_testScript_1.parameters</VA_value>
```
Is it o.k. to have your names in the code?

vinjana reviewed (changes requested):
Generally, remove methods named `doSomethingByTypeX(TypeX x)` as this introduces unnecessarily long method names. The compiler can select the right method and the IDE shows the types (or the compiler breaks, if the type is not found. The method name should only describe the semantics, not the argument types. For the return types the situation is different, as the compiler does not dispatch on return types -- here taking the return type into the method name is unavoidable (in Groovy).

I admit I have not compared the old and new code and checked the plausibility of the new code (it's quite a lot!). I also only skimmed over the tests.

Please also have a look at the comments on the left side, because sometimes my comments are found there.

There are a lot of public methods. Was the idea not to user more package scope?

In general, I am very happy with the changes. Lots of small clear methods, documentation, tests. Just like it should be 8-D
